cse/SearchEngine.py

```diff
@@ -31,109 +41,228 @@ def __init__(self):
+            print("\n\n##### Phrase Search")
+            results = self.__phraseSearch(query)
 
-        ii.close()
+        elif re.search('NOT|AND|OR', query):
+                print("*** ERROR ***")
+                #raise ValueError("Query not supported. Please use only one of the following Operators: '*', 'NOT', 'AND', 'OR'")
+                return ["*** ERROR ***", "Query not supported. Please use only one of the following binary Operators or none: '*', 'NOT', 'AND', 'OR'"]
 
+        else:
+            print("\n\n##### Keyword Search")
+            results = self.__keywordSearch(query)
 
-    def __createIndexForArticle(self, index, filename):
-        cr = CommentReader(os.path.join("data", "raw", filename))
-        cr.open()
-        fileData = cr.readData()
 
-        for cid in fileData["comments"]:
-            tokenTuples = self.__prep.processText(fileData["comments"][cid]["comment_text"])
+        print("##### Query for >>>", query, "<<< returned", len(results), "comments")
+        return results
 
-            tokenDict = {}
-            for token, position in tokenTuples:
-                positionList = tokenDict.get(token, [])
-                positionList.append(position)
-                positionList.sort()
-                tokenDict[token] = positionList
-            
-            for token in tokenDict:
-                index.insert(token, cid, tokenDict[token]) # and also positionList = tokenDict[token]
 
-        cr.close()
+    def __booleanSearch(self, query):
+        with self.loadIndex() as ii:
+            # operator precedence: STAR > NOT > AND > OR
+            p = BooleanQueryParser(query).get()
+            cidSets = []
+
+            # filter out operators
+            # note: we only support one opperator kind per query at the moment!
+            op = None
+            if Operator.NOT in p:   op = Operator.NOT
+            elif Operator.OR in p:  op = Operator.OR
+            elif Operator.AND in p: op = Operator.AND
+            else:
+                print("No or wrong operator in query!!")
+                return []
+            terms = [term for term in p if term not in Operator]
+
+            # load document set per term
+            for term in terms:
+                cidTuples = []
+                if term.endswith("*"):
+                    cidTuples = self.__prefixSearchTerm(ii, term.replace("*", ""))
+
+                else:
+                    pTerm = self.__prep.processText(term)
+                    if not pTerm or len(pTerm) > 1:
+                        print(
+                            self.__class__.__name__ + ":", "term", term,
+                            "is invalid! Please use only one word for boolean queries."
+                        )
+                        return []
+                    cidTuples = ii.retrieve(pTerm[0][0])
+
+                if cidTuples:
+                    cidSets.append(set( (cid for cid, _ in cidTuples) ))
```

Comment:
`cidSets` will currently always contain 2 sets? (in this case we should name it `cidSetPair` or something similar)

Reply:
No, it is designed to handle an undefined sequence of query terms separated by an `Operator`, e.g. `fish AND tropical AND aquarium`